reps/2023-8-18-ray-on-spark-autoscaling.md

```diff
@@ -0,0 +1,230 @@
+
+  ```
+  pip install ray
+  ```
+
+  If the user's workload requires Ray add-ons, they must also be installed, e.g. via
+
+  ```
+  pip install "ray[debug,dashboard,tune,rllib,serve]"
+  ```
+
+### Ray on spark autoscaling interfaces
+
+Integrate autoscaling feature into existing `ray.util.spark.cluster_init.setup_ray_cluster` API,
+The following new arguments are added:
+
+ - autoscale: bool (default False)
+   If True, enable autoscaling, the number of initial Ray worker nodes 
+   is zero, and the maximum number of Ray worker nodes is set to
+   `num_worker_nodes`. Default value is False.
+ - autoscale_upscaling_speed: Optional[float] (default 1.0)
+   The upscaling speed configuration, see explanation in [here](https://docs.ray.io/en/latest/cluster/vms/user-guides/configuring-autoscaling.html#upscaling-and-downscaling-speed).
+ - autoscale_idle_timeout_minutes: Optional[float] (default 1.0)
+   The downscaling speed configuration, see explanation in [here](https://docs.ray.io/en/latest/cluster/vms/user-guides/configuring-autoscaling.html#upscaling-and-downscaling-speed).
+ 
+
+### Implementation
+
+To start a Ray cluster that enables autoscaling, firstly, prepares a config file like following:
+
+```yaml
+cluster_name: spark
+max_workers: 8  # maximum number of workers
+provider:
+    # This is the key of registered Ray autoscaling `NodeProvider` class that is
+    # used as the backend of launching Ray worker nodes.
+    type: spark
+    # This must be true since the nodes share the same ip!
+    use_node_id_as_ip: True
+    disable_node_updaters: True
+    disable_launch_config_check: True
+available_node_types:
+    ray.head.default:  # Ray head configurations
+        resources:
+            CPU: 4
+            GPU: 0
+        node_config: {}
+        max_workers: 0
+    ray.worker:  # Ray worker configurations
+        resources:
+            CPU: 4
+            GPU: 2
+            memory: 2000000000
+            object_store_memory: 1000000000
+        node_config: {}
+        min_workers: 0  # minimum number of workers in this group
+        max_workers: 4  # maximum number of workers in this group
+head_node_type: ray.head.default
+upscaling_speed: 1.0
+idle_timeout_minutes: 0.1
+```
+
+Then, starts a Ray head node via following command (some options are omitted):
+
+```shell
+ray start --head --autoscaling-config=/path/to/autoscaler-config.yaml
+```
+
+The above command starts a Ray head node, it also starts a Ray autoscaler with
+above YAML configuration, it configures autoscaler with "spark" `NodeProvider`
+backend. When Ray cluster is requested to scale up,
+the configured `NodeProvider` is called to trigger Ray worker node setup.
+
+This is `NodeProvider` [interfaces](https://github.com/ray-project/ray/blob/e1c48ab8b1225daa44ea63ed1e7dc956b0e7f411/python/ray/autoscaler/node_provider.py#L13).
+
+### How to implement spark node provider?
+
+For each Ray worker node, each Ray worker node, we start a spark job with only one spark
```

Comment:
I think we should just explain why we choose to have 1:1:1 mapping between spark job, spark task, ray worker node here instead of down below.